SonarQube.TeamBuild.PreProcessor/Roslyn/EmbeddedAnalyzerInstaller.cs

```diff
@@ -0,0 +1,208 @@
+            }
+
+            if (!plugins.Any())
+            {
+                this.logger.LogInfo(RoslynResources.EAI_NoPluginsSpecified);
+                return Enumerable.Empty<string>(); // nothing to deploy
+            }
+
+            this.logger.LogInfo(RoslynResources.EAI_InstallingAnalyzers);
+
+            this.logger.LogDebug(RoslynResources.EAI_LocalAnalyzerCache, this.localCacheDirectory);
+            Directory.CreateDirectory(this.localCacheDirectory); // ensure the cache dir exists
+
+            HashSet<string> allFiles = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+            foreach (Plugin plugin in plugins)
+            {
+                IEnumerable<string> files = GetPluginResourceFiles(plugin);
+                foreach(string file in files)
+                {
+                    allFiles.Add(file);
+                }
+            }
+
+            return allFiles;
+        }
+
+        #endregion
+
+        #region Private methods
+
+        /// <summary>
+        /// We want the resource cache to be in a well-known location so we can re-use files that have
+        /// already been installed (although this won't help for e.g. hosted build agents)
+        /// </summary>
+        private static string GetLocalCacheDirectory()
+        {
+            string localCache = Path.Combine(Path.GetTempPath(), ".sonarqube", ".nuget");
+            return localCache;
+        }
+        
+        private IEnumerable<string> GetPluginResourceFiles(Plugin plugin)
+        {
+            string cacheDir = GetResourceSpecificDir(this.localCacheDirectory, plugin);
+
+            if (Directory.Exists(cacheDir))
+            {
+                this.logger.LogDebug(RoslynResources.EAI_UsingCachedResource, cacheDir);
+            }
+            else
+            {
+                FetchResourceFromServer(plugin, cacheDir);
+            }
+
+            string[] allFiles = Directory.GetFiles(cacheDir, "*.*", SearchOption.AllDirectories);
+            return allFiles;
+        }
+
+        private void FetchResourceFromServer(Plugin plugin, string targetDir)
+        {
+            logger.LogDebug(RoslynResources.EAI_FetchingPluginResource, plugin.Key, plugin.Version, plugin.StaticResourceName);
+
+            Directory.CreateDirectory(targetDir);
+
+            if (server.TryDownloadEmbeddedFile(plugin.Key, plugin.StaticResourceName, targetDir))
+            {
+                string targetFilePath = Path.Combine(targetDir, plugin.StaticResourceName);
+
+                if (IsZipFile(targetFilePath))
+                {
+                    logger.LogDebug(Resources.MSG_ExtractingFiles, targetDir);
+                    ZipFile.ExtractToDirectory(targetFilePath, targetDir);
+                }
+            }
+            else
+            {
+                logger.LogWarning(RoslynResources.EAI_PluginResourceNotFound, plugin.Key, plugin.Version, plugin.StaticResourceName);
```

Comment:
Failing wouldn't be a great user experience. If the file isn't a zip file then we can just leave it as it is.